File: 39. Combination Sum.cpp
class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        DFS({}, candidates, target);
        return res;
    }
    
    void DFS(vector <int> path, vector<int>& candidates, int target)
    {
        if (target == 0)
        {
            res.push_back(path);
            return;
        }
        for (int i = 0; i < candidates.size(); i++)
        {
            if (candidates[i] <= target)
            {
                if (path.size() > 0 && candidates[i] < path[path.size() - 1])
                    continue;
                path.push_back(candidates[i]);
                DFS(path, candidates, target - candidates[i]);
                path.pop_back();
            }
        }
    }
};

File: 39. Combination Sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        self.res = []
        self.DFS([], candidates, target)
        return list(self.res)
    
    def DFS(self, path, candidates, target):
        if target == 0:
            self.res += [path]
            return
        for i in candidates:
            if i <= target:
                if len(path) == 0:
                    self.DFS(path + [i], candidates, target - i)
                elif i >= path[len(path) - 1]:
                    self.DFS(path + [i], candidates, target - i)
